labgrid/driver/rawnetworkinterfacedriver.py

```diff
@@ -0,0 +1,90 @@
+# pylint: disable=no-member
+import json
+import signal
+import subprocess
+import shutil
+import tempfile
+
+import attr
+
+from .common import Driver
+from ..factory import target_factory
+from ..step import step
+from ..util.helper import processwrapper
+from ..util.managedfile import ManagedFile
+from ..resource.common import NetworkResource
+
+
+@target_factory.reg_driver
+@attr.s(eq=False)
+class RawNetworkInterfaceDriver(Driver):
+    bindings = {
+        "iface": {"NetworkInterface", "RemoteNetworkInterface", "USBNetworkInterface"},
+    }
+
+    def __attrs_post_init__(self):
+        super().__attrs_post_init__()
+        self.proc_pcap_map = {}
+
+    def _get_wrapper_prefix(self):
+        return self.iface.command_prefix + ["sudo", "labgrid-raw-interface"]
+
+    @Driver.check_active
+    @step(args=["count"])
+    def record(self, *, count=None):
+        capture = tempfile.NamedTemporaryFile(prefix="lg-raw-int-cap-")
+        cmd = self._get_wrapper_prefix() + ["tcpdump", self.iface.ifname]
+        if count is not None:
+            cmd.append(str(count))
+
+        proc = subprocess.Popen(cmd, stdout=capture, stderr=subprocess.PIPE)
+        self.proc_pcap_map[proc] = capture
+        return proc
+
+    @Driver.check_active
+    @step(args=["proc", "filename"])
+    def get_record(self, proc, filename, *, timeout=None):
+        assert proc in self.proc_pcap_map
+
+        try:
+            _, err = proc.communicate(timeout=timeout)
+        except subprocess.TimeoutExpired:
+            proc.send_signal(signal.SIGINT)
+            _, err = proc.communicate()
+
+        assert proc.returncode == 0, f"tcpdump err={err}"
+
+        capfile = self.proc_pcap_map[proc]
+        shutil.copy(capfile.name, filename)
+        capfile.close()
+
+    @Driver.check_active
+    @step(args=["filename"])
+    def start_replay(self, filename):
+        mf = ManagedFile(filename, self.iface)
+        mf.sync_to_resource()
+
+        if isinstance(self.iface, NetworkResource):
+            cmd = self.iface.command_prefix
+            cmd.append(f"sudo labgrid-raw-interface tcpreplay {self.iface.ifname} < {mf.get_remote_path()}")
+            indata = None
+        else:
+            cmd = self._get_wrapper_prefix() + ["tcpreplay", self.iface.ifname]
+            indata = open(mf.get_remote_path(), "rb").read()
+        return subprocess.run(cmd, input=indata)
```

Comment:
This waits for the completion of the subprocess. When I see the method name `start_replay()`, I expect receiving a subprocess handle instead of a CompletedProcess object. I think this is wrong.

Reply:
The interesting problem here is that with Popen, we can't simply pass in input data as with subprocess run. We probably need to `Popen` with `stdin=PIPE`, pipe in the data and than close `stdin` on the other side.

Reply:
I've integrated the suggested change, please have a look.

Reply:
Can't we just pass the open FD as stdin to the subprocess and then close it on our side?